src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,372 @@
+        {
+            this.chainState = chainState;
+            this.nodeLifetime = nodeLifetime;
+            this.storeSettings = storeSettings;
+            this.chain = chain;
+            this.blockRepository = blockRepository;
+
+            this.blocksQueue = new AsyncQueue<BlockPair>();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <summary>
+        /// Initializes the <see cref="BlockStore"/>.
+        /// <para>
+        /// If <see cref="storeTip"/> is <c>null</c>, the store is out of sync. This can happen when:</para>
+        /// <list>
+        ///     <item>The node crashed.</item>
+        ///     <item>The node was not closed down properly.</item>
+        /// </list>
+        /// <para>
+        /// To recover we walk back the chain until a common block header is found and set the <see cref="BlockStore"/>'s <see cref="storeTip"/> to that.
+        /// </para>
+        /// </summary>
+        public async Task InitializeAsync()
+        {
+            this.logger.LogTrace("()");
+
+            if (this.storeSettings.ReIndex)
+                throw new NotSupportedException();
+
+            ChainedHeader initializationTip = this.chain.GetBlock(this.blockRepository.BlockHash);
+            this.SetStoreTip(initializationTip);
+            
+            if (this.storeTip == null)
+                await this.RecoverStoreTipAsync().ConfigureAwait(false);
+
+            this.logger.LogDebug("Initialized block store tip at '{0}'.", this.storeTip);
+
+            if (this.storeSettings.TxIndex != this.blockRepository.TxIndex)
+            {
+                if (this.storeTip != this.chain.Genesis)
+                    throw new BlockStoreException("You need to rebuild the block store database using -reindex-chainstate to change -txindex");
+
+                if (this.storeSettings.TxIndex)
+                    await this.blockRepository.SetTxIndexAsync(this.storeSettings.TxIndex).ConfigureAwait(false);
+            }
+            
+            // Throw if block store was initialized after the consensus.
+            // This is needed in order to rewind consensus in case it is initialized ahead of the block store.
+            if (this.chainState.ConsensusTip != null)
+            {
+                this.logger.LogCritical("Block store initialized after the consensus!");
+                throw new BlockStoreException("Block store initialized after consensus!");
+            }
+
+            // Start dequeuing.
+            this.currentBatchSizeBytes = 0;
+            this.dequeueLoopTask = this.DequeueBlocksContinuouslyAsync();
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>Sets the internal store tip and exposes the store tip to other components through the chain state.</summary>
+        private void SetStoreTip(ChainedHeader newTip)
+        {
+            this.storeTip = newTip;
+            this.chainState.BlockStoreTip = newTip;
```

Comment:
You know what, I don't really like this kind of constructs.
When you have
```
this.storeTip = newTip;
this.chainState.BlockStoreTip = newTip;
```
the 2 store tips can potentially get out of sync, and then we're in a world of confusion as to where the truth is. Look in the Wallet Reorg bit, I've been arguing with @dangershony for over a year about this lol.

Ideally you'd want to have a single source of truth.
So if you could expose the BlockStore.storeTip and retrieve it in chainState.BlockStoreTip's get, it's better IMO.

Reply:
It's all about exposing some property to other components that don't have block store as a reference while still maintaining the full control over the value. 

In case ChainState.StoreTip is used as the only value for the tip any random component could set the value and that will lead to block store being corrupted. 

We have the same with the ChainState.ConsensusTip. 
There is no way around duplication unless we add solution dependency (or move everything to the base feature).

Reply:
I hate it! ok, ok fair enough 😄